File: Python solutions/Trees/design_add_search_word_DS.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False

class WordDictionary:

    def __init__(self):
        self.root = TrieNode()
        

    def addWord(self, word: str) -> None:

        cur = self.root

        for c in word:
            if c not in cur.children:
                cur.children[c] = TrieNode()
            cur = cur.children[c]
        
        cur.isEnd = True
        

    def search(self, word: str) -> bool:

        node = self.root

        def dfs(j, word, cur) -> bool:

            for i in range(j,len(word)):

                if word[i] == '.':
                    for temp in cur.children:

                        if dfs(i+1,word,cur.children[temp]):
                            return True
                    return False
                else:
                    if word[i] not in cur.children:
                        return False
                    cur = cur.children[word[i]]
            
            return cur.isEnd
        
        return dfs(0,word,node)
        
        

# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: Java solutions/Trees/designAddSearchWordDS.java
class WordDictionary {

    class TrieNode {
        TrieNode[] children;
        boolean isEnd;

        public TrieNode()
        {
            this.children = new TrieNode[26];
            this.isEnd = false;
        }
    }

    TrieNode root;

    public WordDictionary() {
        this.root = new TrieNode();
    }
    
    public void addWord(String word) {

        TrieNode cur = this.root;

        for(char c : word.toCharArray())
        {
            if(cur.children[c-'a'] == null)
            {
                cur.children[c-'a'] = new TrieNode();
            }
            cur = cur.children[c-'a'];
        }

        cur.isEnd = true;
    }
    
    public boolean search(String word) {

        TrieNode node = root;
        return dfs(0, word.toCharArray(), node);
    }

    private boolean dfs(int idx, char[] word, TrieNode cur)
    {
        for(int i = idx; i < word.length; i++)
        {
            if(word[i] == '.')
            {
                for(TrieNode temp: cur.children)
                {
                    if(temp != null && dfs(i+1, word, temp))
                    {
                        return true;
                    }
                }
                return false;
            }
            else
            {
                if(cur.children[word[i]-'a'] == null)
                {
                    return false;
                }
                cur = cur.children[word[i]-'a'];
            }
        }

        return cur.isEnd;
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */